BUG: parse_spec for "em" might be wrong if you use synphot>=0.1.2

`synphot` changed unit handling behavior for total flux input for `GuassianFlux1D` in spacetelescope/synphot_refactor#154 to fix another bug (spacetelescope/synphot_refactor#153). It was an oversight that `parse_spec` here was not updated to match.

For example: `'em(5000, 25, 1, flam)'` should produce Gaussian with a total flux of 1 `erg / s / cm^2` but because of how `total_flux` is passed in here (as a plain float), the assumption that `synphot` would interpret it as `ph / s / cm^2` is now wrong.

I don't remember why I did not pass it in as a `Quantity`. I must gave a good reason a few years ago? :woman_shrugging: :grimacing:

To illustrate, let's pick a test case from #103 below:

```python
from astropy import units as u
from astropy.tests.helper import assert_quantity_allclose  # For later

from synphot import SourceSpectrum, units
from synphot.models import GaussianFlux1D

from stsynphot import spparser

sp1 = spparser.parse_spec('em(5000, 25, 1, flam)')
w = sp1.waveset  # For later

# Pythonic equivalent
f = 1 * (units.FLAM * u.AA)
sp2 = SourceSpectrum(GaussianFlux1D, mean=5000 * u.AA, fwhm=25 * u.AA, total_flux=f)
```

With `synphot` 0.2.1 and `stsynphot` 0.2.1 (without the patch in #103):
```python
>>> sp1.meta['expr']
'em(5000, 25, 2.51706e+11, FLAM)'
>>> assert_quantity_allclose(sp1(w), sp2(w))
...
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 100 / 100 (100%)
Max absolute difference: 2.38075292e+21
Max relative difference: 2.51705828e+11
 x: array([8.872240e+15, 1.455489e+16, 2.363970e+16, 3.801297e+16,
       6.051718e+16, 9.538557e+16, 1.488483e+17, 2.299651e+17,
       3.517523e+17, 5.326835e+17, 7.986536e+17, 1.185509e+18,...
 y: array([3.524845e+04, 5.782502e+04, 9.391795e+04, 1.510214e+05,
       2.404282e+05, 3.789566e+05, 5.913580e+05, 9.136262e+05,
       1.397474e+06, 2.116294e+06, 3.172964e+06, 4.709898e+06,...
>>> assert_quantity_allclose(sp1.integrate(wavelengths=w), sp2.integrate(wavelengths=w))
...
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 6.33557745e+22
Max relative difference: 2.51705828e+11
 x: array(6.335577e+22)
 y: array(2.517056e+11)
```

With the patch in #103 for `stsynphot` but same `synphot` version as above:
```python
>>> sp1.meta['expr']
'em(5000, 25, 1, FLAM)'
>>> assert_quantity_allclose(sp1(w), sp2(w))  # No error
>>> assert_quantity_allclose(sp1.integrate(wavelengths=w), sp2.integrate(wavelengths=w))  # No error
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,13 @@
 0.3.0 (unreleased)
 ==================
 
+0.2.2 (unreleased)
+==================
+
+- Fix total flux unit handling for ``em`` spectrum generated from
+  ``parse_spec``. [#102]
+- Fix ``.meta['expr']`` value for normalized spectrum generated from
+  ``parse_spec``. [#102]
 
 0.2.1 (2019-12-20)
 ==================

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -105,6 +105,7 @@ stages:
       numpyCmd: 'pip install numpy==1.15.4'
       astropyCmd: 'pip install astropy==2.0.12'
       scipyCmd: 'pip install scipy==1.1.0'
+      synphotCmd: 'pip install synphot==0.1.3'
       pythonVersion: '3.6'
 
   - template: azure-templates.yml

--- a/stsynphot/spparser.py
+++ b/stsynphot/spparser.py
@@ -288,9 +288,9 @@ class Interpreter(GenericASTMatcher):
     def _get_names_from_tree_values(args):
         names = []
         for arg in args:
-            try:
+            if hasattr(arg, 'meta') and 'expr' in arg.meta:
                 names.append(arg.meta['expr'])
-            except Exception:
+            else:
                 names.append(str(arg))
         return '(' + ','.join(names) + ')'
 

--- a/stsynphot/spparser.py
+++ b/stsynphot/spparser.py
@@ -18,6 +18,7 @@ See :ref:`stsynphot-parser` for more details and
 """
 # ASTROPY
 from astropy import log
+from astropy import units as u
 
 # SYNPHOT
 from synphot import exceptions as synexceptions
@@ -283,6 +284,16 @@ class Interpreter(GenericASTMatcher):
         except AttributeError:
             pass  # We only care about this for relatively simple constructs.
 
+    @staticmethod
+    def _get_names_from_tree_values(args):
+        names = []
+        for arg in args:
+            try:
+                names.append(arg.meta['expr'])
+            except Exception:
+                names.append(str(arg))
+        return '(' + ','.join(names) + ')'
+
     def p_functioncall(self, tree):
         # Where all the real interpreter action is.
         # Note that things that should only be done at the top level
@@ -294,7 +305,8 @@ class Interpreter(GenericASTMatcher):
             args = tree[2].value
 
         fname = tree[0].value
-        metadata = {'expr': '{0}{1}'.format(fname, tuple(args))}
+        metadata = {'expr': '{0}{1}'.format(
+            fname, self._get_names_from_tree_values(args))}
 
         if fname not in _SYFUNCTIONS:
             log.error('Unknown function: {0}'.format(fname))
@@ -317,7 +329,7 @@ class Interpreter(GenericASTMatcher):
             # Black body
             elif fname == 'bb':
                 tree.value = SourceSpectrum(
-                    BlackBodyNorm1D, temperature=args[0])
+                    BlackBodyNorm1D, temperature=args[0]*u.K)
 
             # Power law
             elif fname == 'pl':
@@ -327,7 +339,7 @@ class Interpreter(GenericASTMatcher):
                 try:
                     fluxunit = units.validate_unit(args[2])
                     tree.value = SourceSpectrum(
-                        PowerLawFlux1D, amplitude=1*fluxunit, x_0=args[0],
+                        PowerLawFlux1D, amplitude=1*fluxunit, x_0=args[0]*u.AA,
                         alpha=-args[1], meta=metadata)
                 except (synexceptions.SynphotError, NotImplementedError) as e:
                     log.error(str(e))
@@ -336,7 +348,7 @@ class Interpreter(GenericASTMatcher):
             # Box throughput
             elif fname == 'box':
                 tree.value = SpectralElement(
-                    Box1D, amplitude=1, x_0=args[0], width=args[1],
+                    Box1D, amplitude=1, x_0=args[0]*u.AA, width=args[1]*u.AA,
                     meta=metadata)
 
             # Source spectrum from file
@@ -346,7 +358,7 @@ class Interpreter(GenericASTMatcher):
 
             # Passband
             elif fname == 'band':
-                tree.value = spectrum.band(tree[2].svalue)
+                tree.value = spectrum.band(tree[2].svalue)  # string value
                 tree.value.meta.update(metadata)
 
             # Gaussian emission line
@@ -354,12 +366,12 @@ class Interpreter(GenericASTMatcher):
                 if args[3] not in _SYFORMS:
                     log.error('Unrecognized unit: {0}'.format(args[3]))
                     self.error(fname)
-                x0 = args[0]
+                x0 = args[0] * u.AA
                 fluxunit = units.validate_unit(args[3])
-                totflux = units.convert_flux(
-                    x0, args[2] * fluxunit, units.PHOTLAM).value
+                totflux = args[2] * (fluxunit * u.AA)
                 tree.value = SourceSpectrum(
-                    GaussianFlux1D, total_flux=totflux, mean=x0, fwhm=args[1])
+                    GaussianFlux1D, total_flux=totflux, mean=x0,
+                    fwhm=args[1]*u.AA)
 
             # Catalog interpolation
             elif fname == 'icat':
@@ -407,7 +419,8 @@ class Interpreter(GenericASTMatcher):
                     tree.value = sp
                     tree.value.z = args[1]
                 else:
-                    tree.value = SourceSpectrum(ConstFlux1D, amplitude=1)
+                    tree.value = SourceSpectrum(
+                        ConstFlux1D, amplitude=1*units.PHOTLAM)
 
                 tree.value.meta.update(metadata)
 

--- a/stsynphot/tests/test_parser.py
+++ b/stsynphot/tests/test_parser.py
@@ -27,9 +27,12 @@ from .. import catalog, exceptions, observationmode, spectrum, spparser
 from ..config import conf
 
 
-def _single_functioncall(input_str, ans_cls, ans_model):
+def _single_functioncall(input_str, ans_cls, ans_model, ans_name):
     sp = spparser.parse_spec(input_str)
     assert isinstance(sp, ans_cls)
+
+    if ans_name:
+        assert sp.meta['expr'] == ans_name
 
     # Do not check composite model
     if ans_model is not None:
@@ -37,42 +40,57 @@ def _single_functioncall(input_str, ans_cls, ans_model):
 
 
 @pytest.mark.parametrize(
-    ('input_str', 'ans_cls', 'ans_model'),
-    [('unit(1, flam)', SourceSpectrum, ConstFlux1D),
-     ('bb(5000)', SourceSpectrum, None),
-     ('pl(5000, 1, flam)', SourceSpectrum, PowerLawFlux1D),
-     ('box(5000, 1)', SpectralElement, Box1D),
-     ('em(5000, 25, 1, flam)', SourceSpectrum, Gaussian1D),
-     ('rn(bb(5000), box(5000, 10), 17, abmag)', SourceSpectrum, None),
-     ('z(null, 0.1)', SourceSpectrum, ConstFlux1D),
-     ('z(em(5000, 25, 1, flam), 0.1)', SourceSpectrum, None)])
-def test_single_functioncall(input_str, ans_cls, ans_model):
+    ('input_str', 'ans_cls', 'ans_model', 'ans_name'),
+    [('unit(1, flam)', SourceSpectrum, ConstFlux1D,
+      'unit(1.0,flam)'),
+     ('bb(5000)', SourceSpectrum, None,
+      'bb(5000.0)'),
+     ('pl(5000, 1, flam)', SourceSpectrum, PowerLawFlux1D,
+      'pl(5000.0,1.0,flam)'),
+     ('box(5000, 1)', SpectralElement, Box1D,
+      'box(5000.0,1.0)'),
+     ('em(5000, 25, 1, flam)', SourceSpectrum, Gaussian1D,
+      'em(5000, 25, 1, FLAM)'),
+     ('rn(bb(5000), box(5000, 10), 17, abmag)', SourceSpectrum, None,
+      'rn(bb(5000.0),box(5000.0,10.0),17.0,abmag)'),
+     ('z(null, 0.1)', SourceSpectrum, ConstFlux1D,
+      'z(null,0.1)'),
+     ('z(em(5000, 25, 1, flam), 0.1)', SourceSpectrum, None,
+      'z(em(5000, 25, 1, FLAM),0.1)')])
+def test_single_functioncall(input_str, ans_cls, ans_model, ans_name):
     """Test parser function calls."""
-    _single_functioncall(input_str, ans_cls, ans_model)
+    _single_functioncall(input_str, ans_cls, ans_model, ans_name)
 
 
 @pytest.mark.remote_data
 @pytest.mark.parametrize(
-    ('input_str', 'ans_cls', 'ans_model'),
-    [('spec(crcalspec$alpha_lyr_stis_007.fits)', SourceSpectrum, Empirical1D),
-     ('band(v)', spectrum.ObservationSpectralElement, Empirical1D),
-     ('icat(k93models, 5000, 0.5, 0)', SourceSpectrum, None),
-     ('ebmvx(0.3, mwavg)', ExtinctionCurve, Empirical1D),
+    ('input_str', 'ans_cls', 'ans_model', 'ans_name'),
+    [('spec(crcalspec$alpha_lyr_stis_007.fits)', SourceSpectrum, Empirical1D,
+      'spec(crcalspec$alpha_lyr_stis_007.fits)'),
+     ('band(v)', spectrum.ObservationSpectralElement, Empirical1D, 'band(v)'),
+     ('icat(k93models, 5000, 0.5, 0)', SourceSpectrum, None,
+      'k93models(T_eff=5000,metallicity=0.5,log_g=0)'),
+     ('ebmvx(0.3, mwavg)', ExtinctionCurve, Empirical1D, 'ebmvx(0.3,mwavg)'),
      ('rn(crcalspec$gd71_mod_005.fits, box(5000, 10), 17, vegamag)',
-      SourceSpectrum, None),
+      SourceSpectrum, None,
+      'rn(crcalspec$gd71_mod_005.fits,box(5000.0,10.0),17.0,vegamag)'),
      ('rn(icat(k93models, 5000, 0.5, 0), '
       'cracscomp$acs_f814w_hrc_006_syn.fits, 17, obmag)',
-      SourceSpectrum, None),
+      SourceSpectrum, None,
+      'rn(k93models(T_eff=5000,metallicity=0.5,log_g=0),'
+      'cracscomp$acs_f814w_hrc_006_syn.fits,17.0,obmag)'),
      ('rn(pl(5000, 1, flam), band(v), 1, photlam)',
-      SourceSpectrum, None),
+      SourceSpectrum, None, 'rn(pl(5000.0,1.0,flam),band(v),1.0,photlam)'),
      ('rn(unit(1,flam), band(acs, wfc1, fr388n#3881.0), 10, abmag)',
-      SourceSpectrum, None),
+      SourceSpectrum, None,
+      'rn(unit(1.0,flam),band(acs,wfc1,fr388n#3881.0),10.0,abmag)'),
      ('rn(crcalspec$bd_75d325_stis_002.fits, band(u), 9.5, vegamag) * '
-      'band(fos, blue, 4.3, g160l)', SourceSpectrum, None),
-     ('z(crcalspec$alpha_lyr_stis_007.fits, 0.1)', SourceSpectrum, None)])
-def test_single_functioncall_remote(input_str, ans_cls, ans_model):
+      'band(fos, blue, 4.3, g160l)', SourceSpectrum, None, ''),
+     ('z(crcalspec$alpha_lyr_stis_007.fits, 0.1)', SourceSpectrum, None,
+      'z(crcalspec$alpha_lyr_stis_007.fits,0.1)')])
+def test_single_functioncall_remote(input_str, ans_cls, ans_model, ans_name):
     """Test parser function calls with remote data."""
-    _single_functioncall(input_str, ans_cls, ans_model)
+    _single_functioncall(input_str, ans_cls, ans_model, ans_name)
 
 
 @pytest.mark.remote_data
@@ -90,6 +108,10 @@ class TestRenormPartialOverlap(object):
             sp = spparser.parse_spec(input_str)
         assert isinstance(sp, SourceSpectrum)
         assert 'force_renorm' in sp.warnings
+
+        name = sp.meta['expr']
+        assert (name.startswith('rn(') and
+                name.endswith('qso_fos_001.dat,band(johnson,u),15.0,abmag)'))
 
     def test_disjoint(self):
         """Raise error."""

--- a/stsynphot/tests/test_parser.py
+++ b/stsynphot/tests/test_parser.py
@@ -13,12 +13,16 @@ import os
 
 # THIRD-PARTY
 import pytest
+from astropy import units as u
+from astropy.tests.helper import assert_quantity_allclose
 from astropy.utils.exceptions import AstropyUserWarning
+from numpy.testing import assert_allclose
 
 # SYNPHOT
 from synphot import exceptions as synexceptions
-from synphot.models import (Box1D, ConstFlux1D, Empirical1D, Gaussian1D,
-                            PowerLawFlux1D)
+from synphot import units
+from synphot.models import (BlackBodyNorm1D, Box1D, ConstFlux1D, Empirical1D,
+                            GaussianFlux1D, PowerLawFlux1D)
 from synphot.reddening import ExtinctionCurve
 from synphot.spectrum import SourceSpectrum, SpectralElement
 
@@ -27,70 +31,242 @@ from .. import catalog, exceptions, observationmode, spectrum, spparser
 from ..config import conf
 
 
-def _single_functioncall(input_str, ans_cls, ans_model, ans_name):
-    sp = spparser.parse_spec(input_str)
+def _single_functioncall(sp, ans_cls, ans_model, ans_name, ans_z=0):
     assert isinstance(sp, ans_cls)
-
-    if ans_name:
-        assert sp.meta['expr'] == ans_name
 
     # Do not check composite model
     if ans_model is not None:
         assert isinstance(sp.model, ans_model)
 
+    if ans_name:
+        assert sp.meta['expr'] == ans_name
+    if ans_z is not None:
+        assert_allclose(sp.z, ans_z)
 
-@pytest.mark.parametrize(
-    ('input_str', 'ans_cls', 'ans_model', 'ans_name'),
-    [('unit(1, flam)', SourceSpectrum, ConstFlux1D,
-      'unit(1.0,flam)'),
-     ('bb(5000)', SourceSpectrum, None,
-      'bb(5000.0)'),
-     ('pl(5000, 1, flam)', SourceSpectrum, PowerLawFlux1D,
-      'pl(5000.0,1.0,flam)'),
-     ('box(5000, 1)', SpectralElement, Box1D,
-      'box(5000.0,1.0)'),
-     ('em(5000, 25, 1, flam)', SourceSpectrum, Gaussian1D,
-      'em(5000, 25, 1, FLAM)'),
-     ('rn(bb(5000), box(5000, 10), 17, abmag)', SourceSpectrum, None,
-      'rn(bb(5000.0),box(5000.0,10.0),17.0,abmag)'),
-     ('z(null, 0.1)', SourceSpectrum, ConstFlux1D,
-      'z(null,0.1)'),
-     ('z(em(5000, 25, 1, flam), 0.1)', SourceSpectrum, None,
-      'z(em(5000, 25, 1, FLAM),0.1)')])
-def test_single_functioncall(input_str, ans_cls, ans_model, ans_name):
-    """Test parser function calls."""
-    _single_functioncall(input_str, ans_cls, ans_model, ans_name)
+
+def _compare_spectra(sp1, sp2):
+    """Test that two spectra are basically equivalent."""
+    if sp1.waveset is None:
+        assert sp2.waveset is None
+        w = [100, 5000, 11000] * u.AA
+    else:
+        w = sp1.waveset
+        assert_quantity_allclose(w, sp2.waveset)
+    assert_quantity_allclose(sp1(w), sp2(w))
+    assert_quantity_allclose(sp1.integrate(wavelengths=w),
+                             sp2.integrate(wavelengths=w))
+    assert type(sp1.model.__class__) == type(sp2.model.__class__)
+    if hasattr(sp1, 'z'):
+        assert sp1.z == sp2.z
+
+
+def test_unit_1_flam():
+    sp1 = spparser.parse_spec('unit(1, flam)')
+    _single_functioncall(sp1, SourceSpectrum, ConstFlux1D, 'unit(1.0,flam)')
+
+    sp2 = SourceSpectrum(ConstFlux1D, amplitude=1 * units.FLAM)
+    _compare_spectra(sp1, sp2)
+
+
+def test_bb_5000():
+    sp1 = spparser.parse_spec('bb(5000)')
+    _single_functioncall(sp1, SourceSpectrum, BlackBodyNorm1D, 'bb(5000.0)')
+
+    sp2 = SourceSpectrum(BlackBodyNorm1D, temperature=5000 * u.K)
+    _compare_spectra(sp1, sp2)
+
+
+def test_powerlaw_5000_1_flam():
+    sp1 = spparser.parse_spec('pl(5000, 1, flam)')
+    _single_functioncall(
+        sp1, SourceSpectrum, PowerLawFlux1D, 'pl(5000.0,1.0,flam)')
+
+    sp2 = SourceSpectrum(PowerLawFlux1D, amplitude=1 * units.FLAM,
+                         x_0=5000 * u.AA, alpha=-1)
+    _compare_spectra(sp1, sp2)
+
+
+def test_box_5000_1():
+    sp1 = spparser.parse_spec('box(5000, 1)')
+    _single_functioncall(sp1, SpectralElement, Box1D, 'box(5000.0,1.0)',
+                         ans_z=None)
+
+    sp2 = SpectralElement(Box1D, amplitude=1, x_0=5000 * u.AA, width=1 * u.AA)
+    _compare_spectra(sp1, sp2)
+
+
+def test_em_5000_25_1_flam():
+    sp1 = spparser.parse_spec('em(5000, 25, 1, flam)')
+    _single_functioncall(
+        sp1, SourceSpectrum, GaussianFlux1D, 'em(5000, 25, 1, FLAM)')
+
+    f = 1 * (units.FLAM * u.AA)  # Integrated flux
+    sp2 = SourceSpectrum(
+        GaussianFlux1D, mean=5000 * u.AA, fwhm=25 * u.AA, total_flux=f)
+    _compare_spectra(sp1, sp2)
+
+
+def test_rn_bb_box_abmag():
+    sp1 = spparser.parse_spec('rn(bb(5000), box(5000, 10), 17, abmag)')
+    _single_functioncall(sp1, SourceSpectrum, None,
+                         'rn(bb(5000.0),box(5000.0,10.0),17.0,abmag)')
+
+    bb = SourceSpectrum(BlackBodyNorm1D, temperature=5000 * u.K)
+    box = SpectralElement(Box1D, amplitude=1, x_0=5000 * u.AA, width=10 * u.AA)
+    sp2 = bb.normalize(17 * u.ABmag, band=box)
+    _compare_spectra(sp1, sp2)
+
+
+def test_z_null():
+    """ETC junk spectrum results in flat spectrum with no redshift."""
+    sp1 = spparser.parse_spec('z(null, 0.1)')
+    _single_functioncall(sp1, SourceSpectrum, ConstFlux1D, 'z(null,0.1)')
+
+    sp2 = SourceSpectrum(ConstFlux1D, amplitude=1 * units.PHOTLAM)
+    _compare_spectra(sp1, sp2)
+
+
+def test_z_em():
+    sp1 = spparser.parse_spec('z(em(5000, 25, 1, flam), 0.1)')
+    _single_functioncall(
+        sp1, SourceSpectrum, None, 'z(em(5000, 25, 1, FLAM),0.1)', ans_z=0.1)
+
+    f = 1 * (units.FLAM * u.AA)  # Integrated flux
+    sp2 = SourceSpectrum(
+        GaussianFlux1D, mean=5000 * u.AA, fwhm=25 * u.AA, total_flux=f)
+    sp2.z = 0.1
+    _compare_spectra(sp1, sp2)
 
 
 @pytest.mark.remote_data
-@pytest.mark.parametrize(
-    ('input_str', 'ans_cls', 'ans_model', 'ans_name'),
-    [('spec(crcalspec$alpha_lyr_stis_007.fits)', SourceSpectrum, Empirical1D,
-      'spec(crcalspec$alpha_lyr_stis_007.fits)'),
-     ('band(v)', spectrum.ObservationSpectralElement, Empirical1D, 'band(v)'),
-     ('icat(k93models, 5000, 0.5, 0)', SourceSpectrum, None,
-      'k93models(T_eff=5000,metallicity=0.5,log_g=0)'),
-     ('ebmvx(0.3, mwavg)', ExtinctionCurve, Empirical1D, 'ebmvx(0.3,mwavg)'),
-     ('rn(crcalspec$gd71_mod_005.fits, box(5000, 10), 17, vegamag)',
-      SourceSpectrum, None,
-      'rn(crcalspec$gd71_mod_005.fits,box(5000.0,10.0),17.0,vegamag)'),
-     ('rn(icat(k93models, 5000, 0.5, 0), '
-      'cracscomp$acs_f814w_hrc_006_syn.fits, 17, obmag)',
-      SourceSpectrum, None,
-      'rn(k93models(T_eff=5000,metallicity=0.5,log_g=0),'
-      'cracscomp$acs_f814w_hrc_006_syn.fits,17.0,obmag)'),
-     ('rn(pl(5000, 1, flam), band(v), 1, photlam)',
-      SourceSpectrum, None, 'rn(pl(5000.0,1.0,flam),band(v),1.0,photlam)'),
-     ('rn(unit(1,flam), band(acs, wfc1, fr388n#3881.0), 10, abmag)',
-      SourceSpectrum, None,
-      'rn(unit(1.0,flam),band(acs,wfc1,fr388n#3881.0),10.0,abmag)'),
-     ('rn(crcalspec$bd_75d325_stis_002.fits, band(u), 9.5, vegamag) * '
-      'band(fos, blue, 4.3, g160l)', SourceSpectrum, None, ''),
-     ('z(crcalspec$alpha_lyr_stis_007.fits, 0.1)', SourceSpectrum, None,
-      'z(crcalspec$alpha_lyr_stis_007.fits,0.1)')])
-def test_single_functioncall_remote(input_str, ans_cls, ans_model, ans_name):
-    """Test parser function calls with remote data."""
-    _single_functioncall(input_str, ans_cls, ans_model, ans_name)
+def test_remote_spec_vegafile():
+    sp1 = spparser.parse_spec('spec(crcalspec$alpha_lyr_stis_007.fits)')
+    _single_functioncall(sp1, SourceSpectrum, Empirical1D,
+                         'spec(crcalspec$alpha_lyr_stis_007.fits)')
+
+    sp2 = SourceSpectrum.from_file(os.path.join(
+        os.environ['PYSYN_CDBS'], 'calspec', 'alpha_lyr_stis_007.fits'))
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_band_v():
+    sp1 = spparser.parse_spec('band(v)')
+    _single_functioncall(
+        sp1, spectrum.ObservationSpectralElement, Empirical1D, 'band(v)',
+        ans_z=None)
+
+    sp2 = SpectralElement.from_filter('johnson_v')
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_icat_k93():
+    sp1 = spparser.parse_spec('icat(k93models, 5000, 0.5, 0)')
+    _single_functioncall(sp1, SourceSpectrum, Empirical1D,
+                         'k93models(T_eff=5000,metallicity=0.5,log_g=0)')
+
+    sp2 = catalog.grid_to_spec('k93models', 5000, 0.5, 0)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_ebmvx_mwavg():
+    sp1 = spparser.parse_spec('ebmvx(0.3, mwavg)')
+    _single_functioncall(
+        sp1, ExtinctionCurve, Empirical1D, 'ebmvx(0.3,mwavg)', ans_z=None)
+
+    sp2 = spectrum.ebmvx('mwavg', 0.3)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_rn_calspec_box():
+    sp1 = spparser.parse_spec(
+        'rn(crcalspec$gd71_mod_005.fits, box(5000, 10), 17, vegamag)')
+    _single_functioncall(
+        sp1, SourceSpectrum, None,
+        'rn(crcalspec$gd71_mod_005.fits,box(5000.0,10.0),17.0,vegamag)')
+
+    gd71 = SourceSpectrum.from_file(os.path.join(
+        os.environ['PYSYN_CDBS'], 'calspec', 'gd71_mod_005.fits'))
+    box = SpectralElement(Box1D, amplitude=1, x_0=5000 * u.AA, width=10 * u.AA)
+    sp2 = gd71.normalize(17 * units.VEGAMAG, band=box, vegaspec=spectrum.Vega)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_rn_icat_k93():
+    sp1 = spparser.parse_spec(
+        'rn(icat(k93models, 5000, 0.5, 0), '
+        'cracscomp$acs_f814w_hrc_006_syn.fits, 17, obmag)')
+    _single_functioncall(
+        sp1, SourceSpectrum, None,
+        'rn(k93models(T_eff=5000,metallicity=0.5,log_g=0),'
+        'cracscomp$acs_f814w_hrc_006_syn.fits,17.0,obmag)')
+
+    k93 = catalog.grid_to_spec('k93models', 5000, 0.5, 0)
+    bp = SpectralElement.from_file(os.path.join(
+        os.environ['PYSYN_CDBS'], 'comp', 'acs', 'acs_f814w_hrc_006_syn.fits'))
+    sp2 = k93.normalize(17 * units.OBMAG, band=bp, area=conf.area)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_rn_powerlaw():
+    sp1 = spparser.parse_spec('rn(pl(5000, 1, flam), band(v), 1, photlam)')
+    _single_functioncall(sp1, SourceSpectrum, None,
+                         'rn(pl(5000.0,1.0,flam),band(v),1.0,photlam)')
+
+    pl = SourceSpectrum(PowerLawFlux1D, amplitude=1 * units.FLAM,
+                        x_0=5000 * u.AA, alpha=-1)
+    bp = SpectralElement.from_filter('johnson_v')
+    sp2 = pl.normalize(1 * units.PHOTLAM, band=bp)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_rn_unit_1_flam():
+    sp1 = spparser.parse_spec(
+        'rn(unit(1,flam), band(acs, wfc1, fr388n#3881.0), 10, abmag)')
+    _single_functioncall(
+        sp1, SourceSpectrum, None,
+        'rn(unit(1.0,flam),band(acs,wfc1,fr388n#3881.0),10.0,abmag)')
+
+    constsp = SourceSpectrum(ConstFlux1D, amplitude=1 * units.FLAM)
+    bp = spectrum.band('acs, wfc1, fr388n#3881.0')
+    sp2 = constsp.normalize(10 * u.ABmag, band=bp)
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_rn_calspec_u():
+    sp1 = spparser.parse_spec(
+        'rn(crcalspec$bd_75d325_stis_002.fits, band(u), 9.5, vegamag) * '
+        'band(fos, blue, 4.3, g160l)')
+    # NOTE: No expr for this combo.
+    _single_functioncall(sp1, SourceSpectrum, None, '')
+
+    bd75 = SourceSpectrum.from_file(os.path.join(
+        os.environ['PYSYN_CDBS'], 'calspec', 'bd_75d325_stis_002.fits'))
+    bp_u = SpectralElement.from_filter('johnson_u')
+    bd75_norm = bd75.normalize(
+        9.5 * units.VEGAMAG, band=bp_u, vegaspec=spectrum.Vega)
+    bp_fos = spectrum.band('fos, blue, 4.3, g160l')
+    sp2 = bd75_norm * bp_fos
+    _compare_spectra(sp1, sp2)
+
+
+@pytest.mark.remote_data
+def test_remote_z_vega():
+    sp1 = spparser.parse_spec('z(crcalspec$alpha_lyr_stis_007.fits, 0.1)')
+    _single_functioncall(sp1, SourceSpectrum, None,
+                         'z(crcalspec$alpha_lyr_stis_007.fits,0.1)', ans_z=0.1)
+
+    sp2 = SourceSpectrum.from_file(os.path.join(
+        os.environ['PYSYN_CDBS'], 'calspec', 'alpha_lyr_stis_007.fits'))
+    sp2.z = 0.1
+    _compare_spectra(sp1, sp2)
 
 
 @pytest.mark.remote_data

